clang/docs/HLSL/RootSignature.rst

```diff
@@ -0,0 +1,426 @@
+      uint32_t ShaderRegister;
+      uint32_t RegisterSpace = 0;
+      uint32_t Num32BitValues;
+    };
+
+    struct RootDescriptor {
+      uint32_t ShaderRegister;
+      uint32_t RegisterSpace = 0;
+      RootDescriptorFlags Flags = RootDescriptorFlags::None;
+    };
+    struct RootParameter {
+      RootParameterType ParameterType;
+      std::variant<RootDescriptorTable, RootConstants, RootDescriptor>
+            Parameter;
+      ShaderVisibility ShaderVisibility = ShaderVisibility::All;
+    };
+
+    struct StaticSamplerDesc {
+      Filter Filter = Filter::ANISOTROPIC;
+      TextureAddressMode AddressU = TextureAddressMode::Wrap;
+      TextureAddressMode AddressV = TextureAddressMode::Wrap;
+      TextureAddressMode AddressW = TextureAddressMode::Wrap;
+      float MipLODBias = 0.f;
+      uint32_t MaxAnisotropy = 16;
+      ComparisonFunc ComparisonFunc = ComparisonFunc::LessEqual;
+      StaticBorderColor BorderColor = StaticBorderColor::OpaqueWhite;
+      float MinLOD = 0.f;
+      float MaxLOD = MaxLOD;
+      uint32_t ShaderRegister;
+      uint32_t RegisterSpace = 0;
+      ShaderVisibility ShaderVisibility = ShaderVisibility::All;
+    };
+
+    struct RootSignatureDesc {
+      std::vector<RootParameter> Parameters;
+      std::vector<StaticSamplerDesc> StaticSamplers;
+      RootSignatureFlags Flags;
+    };
+
+    struct VersionedRootSignatureDesc {
+      RootSignatureVersion Version;
+      RootSignatureDesc Desc;
+    };
+
+    class HLSLRootSignatureAttr : public InheritableAttr {
+    protected:
+      std::string RootSignatureStr;
+      VersionedRootSignatureDesc RootSignature;
+    };
+
+
+.. code-block::
+
+    def HLSLEntryRootSignature: HLSLRootSignatureAttr {
+      let Spellings = [GNU<"RootSignature">];
+      let Subjects = Subjects<[HLSLEntry]>;
+      let LangOpts = [HLSL];
+      let Args = [StringArgument<"InputString">];
+    }
+
+For case compile to a standalone root signature blob, the 
+HLSLRootSignatureAttr will be bind to a fake empty entry.
+
+In clang code generation, the HLSLRootSignatureAttr in AST will be translated 
+into a global variable with struct type to express the layout and metadata to 
+save things like static sampler, root flags, space and NumDescriptors in LLVM IR. 
+The struct type will be look like this:
+
+.. code-block:: llvm
+
+  %struct.TABLE0 = type { target("dx.rs.desc"),
+                          target("dx.rs.sampler")}
+
+  %struct.RS = type { target("dx.rs.rootconstant", 4), 
+                      %struct.TABLE0,
+                      target("dx.rs.rootdescriptor") }
+
+The metadata will be look like this:
+
+.. code-block::
+
+  !1 = !{ data for static sampler } ; Save informations for single static 
+                                    ; sampler
+  !2 = !{!1} ; All static samplers
+  !3 = !{ data for descriptors } ; Save informations for single descriptor 
+  !4 = !{ !3 } ; All descriptors
+  !5 = !{void ()* @main, %struct.RS undef, i32 rootFlags, !2, !4}
```

Comment:
Can you describe what "data for static sampler" and "data for descriptors" will be?

Reply:
Maybe a ConstantSturct for StaticSamplerDesc / DescriptorRange .
Just need to make sure all the information goes into the backend, nothing lost.